### ISSUE TITLE
Code Editor Out of Sync error while trying to edit some code in a cell in Jupyterlab

### ISSUE THREAD

**Issue by gurolsaglam** [jupyterlab/jupyterlab#9465]
Hello,

I used Google Colab to write some code, I downloaded Jupyterlab a couple hours ago and opened the same code in jupyter lab. However, when I try to edit a cell, an error message comes up saying "Code Editor Out of Sync". This is the error that prints on the javascript console.

(I know the "google.colab.patches" is probably not available, I was trying to delete that part, but the error kind of interferes with it.)

{"model":"#show (and save for report) images and show the beginning (first 10) of the flattened image arrays\r\n\r\nfor i in range(0, len(imagesDS t\n ry: t\n      cv2.imshow(imagesDS[]#)\r\n  exep #t:\r\n    #I added this part because I am using google collaboratory and they disabled cv2.imsho#w()\r\n    #it will not be run if not using google collabora#tory\r\n    from google.colab.patches import cv2_i#mshow\r\n    cv2_imshow(imagesDS[i])\r\n  #save the downsampled images for report\r\n  cv2.imwrite(imageNames[i].split(\".\")[0]+\"2.png\", imagesDS[i])\r\n","view":"#show (and save for report) images and show the beginning (first 10) of the flattened image arrays\n\nfor i in range(0, len(imagesDS)\n try: t\n      cv2.imshow(imagesDS[]#)\n  exep #t:\n    #I added this part because I am using google collaboratory and they disabled cv2.imsho#w()\n    #it will not be run if not using google collabora#tory\n    from google.colab.patches import cv2_i#mshow\n    cv2_imshow(imagesDS[i])\n  #save the downsampled images for report\n  cv2.imwrite(imageNames[i].split(\".\")[0]+\"2.png\", imagesDS[i])\n","selections":[{"uuid":"a21f53ac-f30f-4d09-85e2-a076b21eacbd","start":{"line":3,"column":2},"end":{"line":3,"column":2},"style":{"className":"","displayName":"","color":"black"}}],"cursor":{"line":3,"column":2},"lineSep":null,"mode":"text/x-ipython","change":{"from":{"line":3,"ch":1,"sticky":null},"to":{"line":3,"ch":1,"sticky":null},"text":["t"],"origin":"+input","removed":[""]}}

**Comment by jasongrout** [comment]
We seem to have a number of people reporting that files from Colab hit this issue. It seems the problem has to do with the line endings `\r\n` colab (or possibly something else?) is putting into the file, which are windows-style line endings, compared to `\n` that are more normally used in other systems.

**Comment by jasongrout** [comment]
Just curious, can you post the file here, if it does not contain sensitive data? You might have to rename it to have a .txt extension. Or alternatively, can you open the file in a text editor and paste the contents similar to the above code? I'm curious if the file has `\r\n` internally.

**Comment by telamonian** [comment]
@jasongrout You beat me to the punch. I was looking up the related issues:

#7703
#9340

There were a related complaint tied to vscode. I tried reproducing this on a mac, and could not (regardless of line ending settings). So my guess is that on Windows at least some other notebook editors are inserting CLRF-style `\r\n` endings in places that our current handling does not deal with gracefully.

@gurolsaglam I'll second Jason's request for you to dump the raw contents of your notebook here. Since we're having trouble reproducing the issue locally, I think having a original sample of a troubled notebook is the only way we'll be able to fix this.

**Comment by telamonian** [comment]
@gurolsaglam Also, once you've dumped the notebook here, there's an "easy" fix for this:

- open your `.ipynb` file in a text editor
- find/replace all `\r\n` with just `\n`

**Comment by gurolsaglam** [comment]
> Just curious, can you post the file here, if it does not contain sensitive data? You might have to rename it to have a .txt extension. Or alternatively, can you open the file in a text editor and paste the contents similar to the above code? I'm curious if the file has `\r\n` line endings in the source.

Yes I can, here (I copied and pasted each cell to Jupyter Lab one by one, and it worked, this is the other version):
[cs552_hw2.zip](https://github.com/jupyterlab/jupyterlab/files/5691269/cs552_hw2.zip)

**Comment by telamonian** [comment]
grand. Now I'm at least able to repro the bug. Just unzip that notebook, open it as usual in jlab, then try to change the contents of any input cell. A modal will pop up after a second or two without even having to save:

![image](https://user-images.githubusercontent.com/2263641/102132736-02b29400-3e22-11eb-87a0-fc9ca2104b34.png)

**Comment by telamonian** [comment]
okay, I think I have it mostly figured out. No big surprise, there's a number of different line-ending normalization behaviors going on, and at least some of them are clashing. The real problem seems to be that Codemirror will by default always normalize line endings to LF (ie `\n`) when dumping text from a live editor. Here's the problem function in our code:

https://github.com/jupyterlab/jupyterlab/blob/de51a4752617de5609ec9dda0ec3f47ff9692912/packages/codemirror/src/editor.ts#L1066-L1098

So `this._model.value.text` contains the "real" text (w.r.t. how it appears in the on-disk `.ipynb` file) with the `\r\n` line-endings, while `doc.getValue()` will always normalize the line endings to `\n` regardless. From the Codemirror source:

```typescript
        /** Get the current editor content. You can pass it an optional argument to specify the string to be used to separate lines (defaults to "\n"). */
        getValue(seperator?: string): string;
```

So it looks like the solution will be to make `_checkSync` somehow aware of the line ending convention in the current editor, and then pass that in as an arg to `doc.getValue('\n')` or `doc.getValue('\r\n')`

**Comment by telamonian** [comment]
Urgh. Turns out there's more screwed up than just mismatched line-endings in the `(doc.getValue() === this._model.value.text)` comparison. Apparently, the extra `\r` character in `\r\n` vs '\n' isn't accounted for anywhere in the code that propagates changes to `this._model.value`:

https://github.com/jupyterlab/jupyterlab/blob/de51a4752617de5609ec9dda0ec3f47ff9692912/packages/codemirror/src/editor.ts#L927-L947

Since changes are added or removed from `this._model.value` based on position, this means that all changes get shifted. So the value of `this._model.value` gets completely messed up.

So a better fix may be to ensure that all in-memory representations of editor text use `\n` line-endings at all times. Which I guess is part of what @ian-r-rose was trying to do in #5622? So the question becomes, if he was normalizing all endings to `\n` on file load, why is that apparently skipped here? Maybe he didn't apply his CLRF fix code to notebooks?

**Comment by blois** [comment]
I don't think it would be entirely unreasonable for Colab to normalize to just `\n`.

There aren't many cases where we encounter carriage returns and they also caused https://github.com/googlecolab/colabtools/issues/1821.

**Comment by gurolsaglam** [comment]
This has been open for a long time now, should I be closing it? Should it stay as is? I have never opened an issue before, so if somebody could tell me if I need to do something, that would be great.

**Comment by blois** [comment]
I'm looking into changing the behavior of Colab right now- will circle back when I have an answer there. I'm not aware of an intentional change around this, but there was some refactoring in Colab in early December which could have inadvertently made this happen.

**Comment by jasongrout** [comment]
Thank you very much @blois! IIRC, when I looked into it before, the newline character isn't specified in the notebook file spec, so Colab wasn't doing something wrong. Of course, it would help compatibility if things were normalized.

@gurolsaglam - let's leave it open for now. It's still a problem that is reported occasionally, and until either JLab is fixed to handle this case, or Colab is fixed to normalize things, it will still continue to be a problem people have.

**Comment by blois** [comment]
The change for Colab to omit carriage returns on save is now live.

**Comment by jasongrout** [comment]
> The change for Colab to omit carriage returns on save is now live.

Thanks! And thanks for following up here as well.

**Comment by krassowski** [comment]
But it turns out VS Code does it as well: https://github.com/microsoft/vscode-jupyter/issues/4576. I agree with @telamonian's assessment, we should fallback to conversion of line-endings on first read.

**Comment by jasongrout** [comment]
> Which I guess is part of what @ian-r-rose was trying to do in #5622? So the question becomes, if he was normalizing all endings to `\n` on file load, why is that apparently skipped here?

Looking at this a little more closely - I think Ian was normalizing line endings in text files, but the errors here are stemming from line endings in notebook cells, i.e., inside the JSON datastructure of the notebook, not the actual line endings in a notebook file serialized to disk.

**Comment by jasongrout** [comment]
> But it turns out VS Code does it as well: [microsoft/vscode-jupyter#4576](https://github.com/microsoft/vscode-jupyter/issues/4576). I agree with @telamonian's assessment, we should fallback to conversion of line-endings on first read.

I've essentially implemented this at #11141 (normalizing on cell creation, rather than on first notebook read, though).

**Comment by jasongrout** [comment]
The issue in vs code was fixed today in https://github.com/microsoft/vscode/pull/133762.

**Comment by jasongrout** [comment]
This fix has also been released in jlab 3.0.18.
